### ISSUE TITLE
linter: Failed to load external plugin because no external linter was configured. This means the Oxlint binary was executed directly rather than via napi bindings

### ISSUE THREAD

**Issue by dalisoft** [oxc-project/oxc#13739]
### What version of Oxlint are you using?

1.15.0

### What command did you run?

`bun run oxlint`

### What does your `.oxlintrc.json` config file look like?

```jsonc
{
  "extends": ["./packages/eslint-config-node/oxlintrc.json"],
  "rules": {
    "@typescript-eslint/no-var-requires": "off",
    "complexity": ["error", { "max": 7 }],
    "max-lines-per-function": [
      "error",
      { "max": 48, "skipBlankLines": true, "skipComments": true, "IIFEs": true }
    ]
  },
  "ignorePatterns": [".history", ".release-me", "scripts"]
}
```

And the referenced JSON config

```jsonc
{
  "env": {
    "es2015": true,
    "es2020": true,
    "node": true,
    "browser": false,
    "jest": true
  },
  "globals": {},
  "plugins": [
    "@typescript-eslint",
    "promise",
    "optimize-regex",
    "no-secrets",
    "import"
  ],
  "rules": {
    "brace-style": [
      "error",
      "1tbs",
      {
        "allowSingleLine": true
      }
    ],
    "comma-dangle": [
      "error",
      "never"
    ],
    "comma-spacing": [
      "error",
      {
        "before": false,
        "after": true
      }
    ],
    "func-call-spacing": [
      "error",
      "never"
    ],
    "quotes": [
      "error",
      "single",
      {
        "avoidEscape": true
      }
    ],
    "semi": [
      "error",
      "always"
    ],
    "space-before-function-paren": [
      "error",
      {
        "anonymous": "always",
        "named": "never",
        "asyncArrow": "always"
      }
    ],
    "keyword-spacing": [
      "error",
      {
        "before": true,
        "after": true,
        "overrides": {
          "return": {
            "after": true
          },
          "throw": {
            "after": true
          },
          "case": {
            "after": true
          }
        }
      }
    ],
    "lines-between-class-members": [
      "error",
      "always",
      {
        "exceptAfterSingleLine": false
      }
    ],
    "space-infix-ops": "error",
    "object-curly-spacing": [
      "error",
      "always"
    ],
    "space-before-blocks": "error",
    "no-extra-semi": "error",
    "default-param-last": "off",
    "@typescript-eslint/default-param-last": "error",
    "indent": [
      "error",
      2,
      {
        "SwitchCase": 1,
        "flatTernaryExpressions": false,
        "offsetTernaryExpressions": true
      }
    ],
    "@typescript-eslint/indent": [
      "off"
    ],
    "no-array-constructor": "off",
    "@typescript-eslint/no-array-constructor": "error",
    "no-dupe-class-members": "off",
    "@typescript-eslint/no-dupe-class-members": "error",
    "no-empty-function": [
      "error"
    ],
    "@typescript-eslint/no-empty-function": "off",
    "no-extra-parens": "off",
    "@typescript-eslint/no-extra-parens": [
      "off",
      "all",
      {
        "conditionalAssign": true,
        "nestedBinaryExpressions": false,
        "returnAssign": false,
        "ignoreJSX": "all",
        "enforceForArrowConditionals": false
      }
    ],
    "no-loss-of-precision": "off",
    "@typescript-eslint/no-loss-of-precision": "error",
    "no-loop-func": "off",
    "@typescript-eslint/no-loop-func": "error",
    "no-magic-numbers": "off",
    "@typescript-eslint/no-magic-numbers": [
      "off",
      {
        "ignore": [],
        "ignoreArrayIndexes": true,
        "enforceConst": true,
        "detectObjects": false
      }
    ],
    "no-redeclare": "off",
    "@typescript-eslint/no-redeclare": "error",
    "no-shadow": [
      "error",
      {
        "hoist": "functions"
      }
    ],
    "@typescript-eslint/no-shadow": "off",
    "no-unused-expressions": "off",
    "@typescript-eslint/no-unused-expressions": [
      "error",
      {
        "allowShortCircuit": false,
        "allowTernary": false,
        "allowTaggedTemplates": false
      }
    ],
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": [
      "error",
      {
        "vars": "all",
        "args": "all",
        "argsIgnorePattern": "^_",
        "caughtErrors": "all",
        "caughtErrorsIgnorePattern": "^_",
        "destructuredArrayIgnorePattern": "^_",
        "varsIgnorePattern": "^_",
        "ignoreRestSiblings": false
      }
    ],
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": [
      "error",
      {
        "functions": true,
        "classes": true,
        "variables": true
      }
    ],
    "no-useless-constructor": "off",
    "@typescript-eslint/no-useless-constructor": "error",
    "require-await": "off",
    "@typescript-eslint/require-await": "off",
    "@typescript-eslint/ban-ts-comment": "error",
    "import/extensions": [
      "error",
      "ignorePackages",
      {
        "js": "always",
        "cjs": "never",
        "mjs": "always",
        "jsx": "never",
        "ts": "never",
        "tsx": "never",
        "json": "always",
        "node": "always",
        "wasm": "always"
      }
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "optionalDependencies": false,
        "devDependencies": [
          "test/**",
          "tests/**",
          "spec/**",
          "**/__tests__/**",
          "**/__mocks__/**",
          "test.{js,jsx}",
          "test.{ts,tsx}",
          "test-*.{js,jsx}",
          "test-*.{ts,tsx}",
          "**/*{.,_}{test,spec}.{js,jsx}",
          "**/*{.,_}{test,spec}.{ts,tsx}",
          "**/jest.config.js",
          "**/jest.config.ts",
          "**/jest.setup.js",
          "**/jest.setup.ts",
          "**/vue.config.js",
          "**/vue.config.ts",
          "**/webpack.config.js",
          "**/webpack.config.ts",
          "**/webpack.config.*.js",
          "**/webpack.config.*.ts",
          "**/rollup.config.js",
          "**/rollup.config.ts",
          "**/rollup.config.*.js",
          "**/rollup.config.*.ts",
          "**/gulpfile.js",
          "**/gulpfile.ts",
          "**/gulpfile.*.js",
          "**/gulpfile.*.ts",
          "**/Gruntfile{,.js}",
          "**/Gruntfile{,.ts}",
          "**/protractor.conf.js",
          "**/protractor.conf.ts",
          "**/protractor.conf.*.js",
          "**/protractor.conf.*.ts",
          "**/karma.conf.js",
          "**/karma.conf.ts",
          "**/.eslintrc.js",
          "**/.eslintrc.ts",
          "**/eslint.config.js",
          "**/eslint.config.ts",
          "**/eslint.config.cjs",
          "**/eslint.config.mjs"
        ]
      }
    ],
    "@eslint-community/eslint-comments/no-unused-disable": "error",
    "@eslint-community/eslint-comments/disable-enable-pair": "warn",
    "@typescript-eslint/explicit-function-return-type": "off",
    "@typescript-eslint/explicit-module-boundary-types": "off",
    "@typescript-eslint/class-literal-property-style": [
      "error"
    ],
    "@typescript-eslint/no-non-null-asserted-optional-chain": [
      "error"
    ],
    "operator-linebreak": [
      "off",
      "after"
    ],
    "prefer-template": "error",
    "complexity": [
      "error",
      {
        "max": 7
      }
    ],
    "max-depth": [
      "error",
      {
        "max": 3
      }
    ],
    "max-nested-callbacks": [
      "error",
      {
        "max": 3
      }
    ],
    "max-lines-per-function": [
      "error",
      {
        "max": 60,
        "skipBlankLines": true,
        "skipComments": true,
        "IIFEs": true
      }
    ],
    "max-lines": [
      "error",
      {
        "max": 100,
        "skipBlankLines": true,
        "skipComments": true
      }
    ],
    "no-useless-concat": "error",
    "no-console": "warn",
    "no-template-curly-in-string": "error",
    "no-underscore-dangle": "off",
    "camelcase": "off",
    "import/prefer-default-export": "warn",
    "import/no-mutable-exports": "error",
    "no-secrets/no-secrets": "error",
    "import/no-unresolved": "error",
    "n/no-unsupported-features/es-syntax": [
      "error",
      {
        "ignores": [
          "modules"
        ]
      }
    ]
  },
  "settings": {
    "node": {
      "tryExtensions": [
        ".ts",
        ".js",
        ".d.ts",
        ".html",
        ".md",
        ".json",
        ".wasm",
        ".node"
      ]
    },
    "import/resolver": {
      "node": {
        "extensions": [
          ".ts",
          ".mjs",
          ".js",
          ".d.ts",
          ".html",
          ".md",
          ".json",
          ".wasm",
          ".node"
        ]
      },
      "typescript": {
        "alwaysTryTypes": true,
        "project": [
          "packages/*/tsconfig.json",
          "tsconfig.json"
        ]
      }
    },
    "import/extensions": [
      ".ts",
      ".mjs",
      ".js",
      ".d.ts",
      ".html",
      ".md",
      ".json",
      ".wasm",
      ".node"
    ],
    "import/external-module-folders": [
      "node_modules",
      "node_modules/@types"
    ],
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
        ".d.ts"
      ]
    }
  }
}
```

### What happened?

```bash
❯ bun run oxlint
Failed to parse configuration file.

  × Failed to load external plugin because no external linter was configured. This means the Oxlint binary was executed directly rather than via napi bindings.

error: "oxlint" exited with code 1
```

### Related issues

- https://github.com/oxc-project/oxc/issues/12168 (same error message)
- https://github.com/oxc-project/oxc/pull/12324

### Environment details

- OS: macOS 15.6.1 (24G90)
- Arch: Apple Silicon
- Runtime: Bun

**Comment by overlookmotel** [comment]
Sorry for the not very helpful error message. It was improved in #13699. In next release, the error will be:

```
JS plugins are not supported without `--experimental-js-plugins` CLI option
```

The problem is that we are part-way through merging support for JS plugins.

The way you add a JS plugin is to add its path to `plugins` in config. e.g.:

```json
{
  "plugins": ["oxc", "typescript", "../path/to/my-custom-plugin.js"]
}
```

The first 2 are Oxc's built-in plugins, the 3rd is a custom JS plugin.

In your referenced `eslint-config-node/oxlintrc.json`, you have various plugins listed e.g. `optimize-regex`. Because they're not recognised as built-in Oxc plugins, it's assumed they're custom JS plugins. But JS plugin support is not enabled, so there's an error.

TLDR... you need to remove `optimize-regex` etc from your config.

Note: Support for JS plugins is not yet complete, so `oxlint --experimental-js-plugins` won't run these plugins (yet!)

Closing this because I think I've answered the problem. If you have any comments about JS plugins, would be very happy to hear your thoughts on #9905.

**Comment by dalisoft** [comment]
@overlookmotel Thank you for clarifying. I didn't want add noise to #9905 so i'll add here.

```bash
OS-airlight-js on  master [$✘!?⇡] via 🥟 v1.2.21 on ☁️   took 11s 
❯ bun run oxlint --experimental-js-plugins

thread 'main' panicked at apps/oxlint/src/lib.rs:60:9:
JS plugins are not supported at present
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
OS-airlight-js on  master [$✘!?⇡] via 🥟 v1.2.21 on ☁️   
❯ bun run oxlint --experimental-js-plugins

thread 'main' panicked at apps/oxlint/src/lib.rs:60:9:
JS plugins are not supported at present
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
OS-airlight-js on  master [$✘!?⇡] via 🥟 v1.2.21 on ☁️   
❯ RUST_BACKTRACE=1 bun run oxlint --experimental-js-plugins

thread 'main' panicked at apps/oxlint/src/lib.rs:60:9:
JS plugins are not supported at present
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
OS-airlight-js on  master [$✘!?⇡] via 🥟 v1.2.21 on ☁️   
❯ RUST_BACKTRACE=full bun run oxlint --experimental-js-plugins

thread 'main' panicked at apps/oxlint/src/lib.rs:60:9:
JS plugins are not supported at present
stack backtrace:
   0:        0x1031c9fe8 - __mh_execute_header
   1:        0x102d5972c - __mh_execute_header
   2:        0x1031c8838 - __mh_execute_header
   3:        0x1031c9ea8 - __mh_execute_header
   4:        0x1031c844c - __mh_execute_header
   5:        0x1031f2538 - __mh_execute_header
   6:        0x1031f24d0 - __mh_execute_header
   7:        0x1031f32ec - __mh_execute_header
   8:        0x10321a9e8 - __mh_execute_header
   9:        0x103106358 - __mh_execute_header
  10:        0x102d24940 - __mh_execute_header
  11:        0x102d2491c - __mh_execute_header
  12:        0x102d24ca4 - __mh_execute_header
OS-airlight-js on  master [$✘!?⇡] via 🥟 v1.2.21 on ☁️   
❯ 
```

**Comment by overlookmotel** [comment]
Thanks for testing it. JS plugin support is currently only implemented in a separate project (`napi/oxlint`). #13723 is the PR which will merge support into Oxlint proper. After that's merged, these panics will disappear.

However, please note that we will still have a bit more work to do before JS plugins are ready for use (hence why the `--experimental-js-plugins` CLI option is currently hidden in `oxc --help`). We're on track to finish up the last missing pieces in next 2 weeks, and will make an announcement when they're ready to test out.

Apologies for the cryptic errors in the meantime.

**Comment by dalisoft** [comment]
Those settings aren't critical for me so that's okay. 
I just report the error message for project improvement. 

For testing i am there to help project

**Comment by camc314** [comment]
To add, to help resolve the error you're seeing. In your config file you specify:

> ```
>   "plugins": [
>     "@typescript-eslint",
>     "promise",
>     "optimize-regex",
>     "no-secrets",
>     "import"
>   ],
> ```

oxlint does not support `optimize-regex`,`no-secrets` plugins at present, so it's looking elsewhere for these plugins. If you remove them from the array, the error should dissappear.
